### PR TITLE
[#350] [Enhancement] author-config.csv: make repo's location optional field 

### LINKED ISSUE #350: author-config.csv: make repo's location and branch name optional field
If the field is empty, we can assume that the authors are meant to be assigned to all the repos in repo-config.csv for analysis.

This change can be useful for #18

### PR DESCRIPTION
**Pull request by fzdy1914**
Fixes #350 

As mentioned in #350, now the location field is mandatory for every author.

However, an author may belong to every repo/branch and it is not convenient 
for the user to type it out all.

Let's add support for empty location field in author-config, meaning that the 
user will add this author to all the repo in repo-config.csv

Things to notice, if an author with empty location exist, all the repos will not 
use the default authors even there is no author listed in author-config.csv because
this author with empty location will be auto added to the repo.

### CODE REVIEW

**Comment by fzdy1914** [comment]
@reposense/devs Ready for review

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 1]
```diff
@@ -78,7 +78,11 @@ public RepoConfiguration(String location, String branch) throws InvalidLocationE
     public RepoConfiguration(String location, String branch, List<String> formats, List<String> ignoreGlobList,
             boolean isStandaloneConfigIgnored, List<String> ignoreCommitList) throws InvalidLocationException {
         this.location = location;
-        this.branch = branch;
+        if (!location.isEmpty()) {
+            this.branch = branch;
+        } else {
+            this.branch = "";
+        }
```
Comment (yong24s) [inline src/main/java/reposense/model/RepoConfiguration.java]: Can use ternary expressions to improve readability.

`this.branch = (!location.isEmpty()) ? branch : "";`

File: src/test/java/reposense/parser/RepoConfigurationTest.java [hunk 2]
```diff
@@ -216,4 +216,22 @@ public void repoConfig_withoutFormatsAndCliFormats_useDefaultFormats() throws Pa
         Assert.assertEquals(1, actualConfigs.size());
         Assert.assertEquals(ArgsParser.DEFAULT_FORMATS, actualConfigs.get(0).getFormats());
     }
+
+    @Test
+    public void equal() throws InvalidLocationException {
```
Comment (yong24s) [inline src/test/java/reposense/parser/RepoConfigurationTest.java]: Please use `three part format` for test methods.
`i.e.` featureUnderTest_testScenario_expectedBehavior()

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 3]
```diff
@@ -113,19 +117,35 @@ public static void setDatesToRepoConfigs(
...
+                        logger.warning(String.format(
+                                "Skipping author as %s already in repository %s",
+                                author.getGitId(), repoConfigToAdd.getDisplayName()));
+                        continue;
+                    }
+                    repoConfigToAdd.addAuthor(author);
+                    repoConfigToAdd.addAuthorAliases(author, author.getAuthorAliases());
+                    repoConfigToAdd.setAuthorDisplayName(author, authorConfig.getAuthorDisplayNameMap().get(author));
+                }
+            }
```
Comment (yong24s) [inline src/main/java/reposense/model/RepoConfiguration.java]: I believe this method can be further simplified. 

`i.e.` 
1. Keep current code but track a list of `authorConfigs` with no locations
1. After the for loop in current code,  add those `authorConfigs`. This should reduce the number of loops.
1. Update methods like `addAuthor`, `addAuthorAliases`, ... to prevent duplicates.
1. Apply SLAP - `https://se-edu.github.io/se-book/codeQuality/maximiseReadability/intermediate/slapHard/index.html`.

Comment (fzdy1914) [reply src/main/java/reposense/model/RepoConfiguration.java]: OK, I will try to find a way to simplify it

File: src/test/java/reposense/parser/RepoConfigurationTest.java (hunk 2 above)
Comment (fzdy1914) [reply src/test/java/reposense/parser/RepoConfigurationTest.java]: can you suggest a name? cause it tests the equal() method

Comment (yong24s) [reply src/test/java/reposense/parser/RepoConfigurationTest.java]: Technically you are using `equals()` for testing and not testing it.

From the assertion, I believe you are testing for `repoConfig_sameLocationDifferentBranch_notSame`?

```
Assert.assertNotEquals(emptyLocationEmptyBranchRepoConfig, validLocationValidBranchRepoConfig);
Assert.assertNotEquals(validLocationDefaultBranchRepoConfig, validLocationValidBranchRepoConfig);
```
However, I am not sure why empty RepoConfigs with empty location are equal since it compares with `Objects.hash(location, branch);`

```
Assert.assertEquals(emptyLocationDefaultBranchRepoConfig, emptyLocationEmptyBranchRepoConfig);
Assert.assertEquals(emptyLocationWithBranchRepoConfig, emptyLocationEmptyBranchRepoConfig);
```

Comment (yong24s) [reply src/test/java/reposense/parser/RepoConfigurationTest.java]: Updated the comment above.

Comment (fzdy1914) [reply src/test/java/reposense/parser/RepoConfigurationTest.java]: Thanks for your advice

**Comment by fzdy1914** [comment]
@reposense/devs Ready for review

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 4]
```diff
@@ -284,6 +297,18 @@ public void setAuthorList(List<Author> authorList) {
         });
     }
 
+    public void addAuthorList(List<Author> authorList) {
```
Comment (yamidark) [inline src/main/java/reposense/model/RepoConfiguration.java]: Let's rename this to `addAuthors`.
The current name sounds like it is adding a `List<Author>` into a `Collections<List<Author>>`, when we are actually appending all the authors in the `authorList` to the `RepoConfiguration` own author list.
Also, add header comments for the method.

File: src/test/java/reposense/parser/RepoConfigurationTest.java [hunk 5]
```diff
@@ -216,4 +216,22 @@ public void repoConfig_withoutFormatsAndCliFormats_useDefaultFormats() throws Pa
         Assert.assertEquals(1, actualConfigs.size());
         Assert.assertEquals(ArgsParser.DEFAULT_FORMATS, actualConfigs.get(0).getFormats());
     }
+
+    @Test
+    public void repoConfig_emptyLocation_equal() throws InvalidLocationException {
```
Comment (yamidark) [inline src/test/java/reposense/parser/RepoConfigurationTest.java]: this should be `repoConfig_emptyLocationDifferentBranch_equal()`?

File: src/test/java/reposense/parser/RepoConfigurationTest.java [hunk 6]
```diff
@@ -216,4 +216,22 @@ public void repoConfig_withoutFormatsAndCliFormats_useDefaultFormats() throws Pa
...
+        RepoConfiguration emptyLocationEmptyBranchRepoConfig = new RepoConfiguration("", "");
+        RepoConfiguration emptyLocationDefaultBranchRepoConfig = new RepoConfiguration("");
+        RepoConfiguration emptyLocationWithBranchRepoConfig = new RepoConfiguration("", "master");
+
+        Assert.assertEquals(emptyLocationDefaultBranchRepoConfig, emptyLocationEmptyBranchRepoConfig);
+        Assert.assertEquals(emptyLocationWithBranchRepoConfig, emptyLocationEmptyBranchRepoConfig);
+    }
+
+    @Test
+    public void repoConfig_differentBranch_notEqual() throws InvalidLocationException {
```
Comment (yamidark) [inline src/test/java/reposense/parser/RepoConfigurationTest.java]: this should be `repoConfig_sameLocationDifferentbranch_notEquals()`?

File: src/test/java/reposense/parser/CsvParserTest.java [hunk 7]
```diff
@@ -85,6 +92,24 @@ public void authorConfig_noSpecialCharacter_success() throws IOException {
         Assert.assertEquals(AUTHOR_CONFIG_NO_SPECIAL_CHARACTER_AUTHORS, config.getAuthorList());
     }
 
+    @Test
+    public void authorConfig_emptyLocation_success() throws ParseException, IOException {
+        RepoConfiguration expectedConfig = new RepoConfiguration("");
+
+        String input = String.format("-config %s", TEST_EMPTY_LOCATION_CONFIG_FOLDER);
+        CliArguments cliArguments = ArgsParser.parse(translateCommandline(input));
```
Comment (yamidark) [inline src/test/java/reposense/parser/CsvParserTest.java]: This is a `CsvParser` unit test, no need to bring in the `ArgsParser` component for testing.

Comment (fzdy1914) [reply src/test/java/reposense/parser/CsvParserTest.java]: Referring to the previous method `merge_twoRepoConfigs_success`,  it also uses the `ArgsParser`.  It is used for the constructor of  RepoConfigCsvParser.

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 8]
```diff
@@ -78,19 +78,22 @@ public static void setDatesToRepoConfigs(
...
-
-            if (index == -1) {
-                logger.warning(String.format(
-                        "Repository %s is not found in repo-config.csv.", authorConfig.getLocation()));
-                continue;
+            if (authorConfig.location.isEmpty()) {
+                for (RepoConfiguration repoConfig : repoConfigs) {
+                    repoConfig.addAuthors(authorConfig.getAuthorList());
+                }
+            } else {
```
Comment (yong24s) [inline src/main/java/reposense/model/RepoConfiguration.java]: *Make the Happy Path Prominent* 
`https://se-edu.github.io/se-book/codeQuality/`

Try something like this?
```
int index = repoConfigs.indexOf(authorConfig);

if (!authorConfig.location.isEmpty() && index == -1) {
    logger.warning(String.format(
            "Repository %s is not found in repo-config.csv.", authorConfig.getLocation()));
    continue;
}


if (authorConfig.location.isEmpty()) {
    for (RepoConfiguration repoConfig : repoConfigs) {
        repoConfig.addAuthors(authorConfig.getAuthorList());
    }

    continue;
}

if (index != -1) {
    RepoConfiguration repoConfig = repoConfigs.get(index);
    repoConfig.addAuthors(authorConfig.getAuthorList());
}
```

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 9]
```diff
@@ -78,19 +78,22 @@ public static void setDatesToRepoConfigs(
...
+                int index = repoConfigs.indexOf(authorConfig);
+
+                if (index == -1) {
+                    logger.warning(String.format(
+                            "Repository %s is not found in repo-config.csv.", authorConfig.getLocation()));
+                    continue;
+                }
+
+                RepoConfiguration repoConfigToAdd = repoConfigs.get(index);
```
Comment (yong24s) [inline src/main/java/reposense/model/RepoConfiguration.java]: `repoConfigToAdd` is confusing. 

Keep it simple, `repoConfigToAdd` -> `repoConfig`?

File: src/main/java/reposense/model/RepoLocation.java [hunk 10]
```diff
@@ -41,6 +41,10 @@ public RepoLocation(String location) throws InvalidLocationException {
         }
     }
 
+    public boolean isEmpty() {
```
Comment (yong24s) [inline src/main/java/reposense/model/RepoLocation.java]: `isEmpty()` is too vague. It gives the impression that it is testing `location` or `RepoLocation`'s attributes, but in actual only `repoName` is being tested.

File: src/test/java/reposense/commits/CommitInfoAnalyzerTest.java [hunk 11]
```diff
@@ -18,6 +18,7 @@
 
     @Test
     public void analyzeCommits_allAuthorNoIgnoredCommitsNoDateRange_success() {
+        config.getAuthorAliasMap().clear();
```
Comment (yong24s) [inline src/test/java/reposense/commits/CommitInfoAnalyzerTest.java]: Can elaborate why it is `clear()` is needed now?

File: src/test/java/reposense/git/GitLogTest.java [hunk 12]
```diff
@@ -94,8 +94,8 @@ public void gitLog_sinceDateInFuture_noContent() {
         Assert.assertTrue(content.isEmpty());
 
         date = TestUtil.getDate(1950, Calendar.JANUARY, 1);
-        config.setUntilDate(date);
-        config.setSinceDate(null);
+        config.setSinceDate(date);
+        config.setUntilDate(null);
```
Comment (yong24s) [inline src/test/java/reposense/git/GitLogTest.java]: Is this change part of the PR?

File: src/test/java/reposense/parser/CsvParserTest.java [hunk 13]
```diff
@@ -139,6 +164,65 @@ public void merge_twoRepoConfigs_success() throws ParseException, IOException {
         Assert.assertEquals(REPO_LEVEL_GLOB_LIST, actualConfigs.get(0).getIgnoreGlobList());
     }
 
+    @Test
+    public void merge_emptyLocation_success() throws ParseException, IOException {
+        FIRST_AUTHOR.setIgnoreGlobList(FIRST_AUTHOR_GLOB_LIST);
+        SECOND_AUTHOR.setIgnoreGlobList(REPO_LEVEL_GLOB_LIST);
+
+        List<Author> expectedAuthorsBeta = new ArrayList<>();
+        expectedAuthorsBeta.add(FIRST_AUTHOR);
```
Comment (yong24s) [inline src/test/java/reposense/parser/CsvParserTest.java]: *Can improve code quality using technique: use nouns for things and verbs for actions*
`https://se-edu.github.io/se-book/codeQuality/`

`expectedAuthorsBeta` -> `expectedBetaAuthors`

File: src/test/java/reposense/parser/CsvParserTest.java [hunk 14]
```diff
@@ -139,6 +164,65 @@ public void merge_twoRepoConfigs_success() throws ParseException, IOException {
...
+
+        RepoConfiguration expectedConfigBeta =
+                new RepoConfiguration(new RepoLocation(TEST_REPO_BETA_LOCATION), TEST_REPO_BETA_BRANCH);
+        expectedConfigBeta.setAuthorList(expectedAuthorsBeta);
+        expectedConfigBeta.setAuthorDisplayName(FIRST_AUTHOR, "Nbr");
+        expectedConfigBeta.setAuthorDisplayName(SECOND_AUTHOR, "Zac");
+        expectedConfigBeta.addAuthorAliases(SECOND_AUTHOR,  Arrays.asList("Zachary Tang"));
+        expectedConfigBeta.setIgnoreGlobList(REPO_LEVEL_GLOB_LIST);
+
+        RepoConfiguration expectedConfigDelta =
```
Comment (yong24s) [inline src/test/java/reposense/parser/CsvParserTest.java]: Same as above.

`expectedConfigDelta` -> `expectedDeltaConfig`

File: src/main/java/reposense/model/RepoLocation.java (hunk 10 above)
Comment (fzdy1914) [reply src/main/java/reposense/model/RepoLocation.java]: This is a bug. I have fixed it.

File: src/test/java/reposense/git/GitLogTest.java (hunk 12 above)
Comment (yong24s) [reply src/test/java/reposense/git/GitLogTest.java]: We want to keep this PR with only the required changes to make repo location an optional field.

Also, it is quicker for us to verify and merge the fix, if it is in a separate PR, so other contributors can pull. :-)

**Comment by yamidark** [comment]
@fzdy1914 Do update your branch, and let us know when you are ready for the next iteration of review by tagging @reposense/devs

**Comment by fzdy1914** [comment]
@reposense/devs Ready for review

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 15]
```diff
@@ -48,7 +48,7 @@ public RepoConfiguration(RepoLocation location, String branch) {
     public RepoConfiguration(RepoLocation location, String branch, List<Format> formats, List<String> ignoreGlobList,
             boolean isStandaloneConfigIgnored, List<CommitHash> ignoreCommitList) {
         this.location = location;
-        this.branch = branch;
+        this.branch = location.isEmpty() ? "" : branch;
```
Comment (yamidark) [inline src/main/java/reposense/model/RepoConfiguration.java]: Why is this necessary?

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 16]
```diff
@@ -228,8 +232,21 @@ public void setIgnoreCommitList(List<CommitHash> ignoreCommitList) {
         return authorList;
     }
 
+    public void setAuthorDetails(Author author) {
```
Comment (yamidark) [inline src/main/java/reposense/model/RepoConfiguration.java]: Header comments for this method

File: src/main/java/reposense/model/RepoConfiguration.java (hunk 15 above)
Comment (fzdy1914) [reply src/main/java/reposense/model/RepoConfiguration.java]: I believe this is a kind of defensive programming because a `RepoConfiguration` with empty location do not need to have a branch. If an author config contains a line that has an empty location and non-empty branch, the equal() method of `RepoConfiguration` will fail and the author will not be added to the existing `AuthorConfig` that with empty location. The alternative way is to change the `equal()` method. But I believe doing this way is better.

File: src/main/java/reposense/model/RepoConfiguration.java (hunk 16 above)
Comment (fzdy1914) [reply src/main/java/reposense/model/RepoConfiguration.java]: Can I know what the header comment means? Do you mean javadoc for this method? Or a simple line of word explaining what this method do?

Comment (yamidark) [reply src/main/java/reposense/model/RepoConfiguration.java]: Yes javadocs

**Comment by fzdy1914** [comment]
@reposense/devs Ready for review

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 17]
```diff
@@ -236,28 +240,61 @@ public void setIgnoreCommitList(List<CommitHash> ignoreCommitList) {
...
-            setAuthorDisplayName(author, author.getDisplayName());
-
-            // Propagate RepoConfiguration IgnoreGlobList to Author
-            author.appendIgnoreGlobList(this.getIgnoreGlobList());
+            setAuthorDetails(author);
+            propagateIgnoreGlobList(author);
         });
     }
 
+    public void addAuthors(List<Author> authorList) {
```
Comment (yong24s) [inline src/main/java/reposense/model/RepoConfiguration.java]: Group `addAuthor()` and `addAuthors()` together by moving `addAuthors()` under `addAuthor()`.

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 18]
```diff
@@ -236,28 +240,61 @@ public void setIgnoreCommitList(List<CommitHash> ignoreCommitList) {
...
+        for (Author author: authorList) {
```
Comment (yong24s) [inline src/main/java/reposense/model/RepoConfiguration.java]: Missing a space before `:`

-> `for (Author author : authorList) {`

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 19]
```diff
@@ -236,28 +240,61 @@ public void setIgnoreCommitList(List<CommitHash> ignoreCommitList) {
...
+            if (containsAuthor(author)) {
+                logger.warning(String.format("Skipping author as %s already in repository %s",
+                        author.getGitId(), getDisplayName()));
+                continue;
+            }
+            addAuthor(author);
```
Comment (yong24s) [inline src/main/java/reposense/model/RepoConfiguration.java]: Add a empty line before this.

We practice adding empty lines after block like `if-else`, `for`, `try-catch`

File: src/test/java/reposense/parser/CsvParserTest.java [hunk 20]
```diff
@@ -88,6 +95,22 @@ public void authorConfig_noSpecialCharacter_success() throws IOException, Invali
...
+
+        AuthorConfigCsvParser authorConfigCsvParser =
+                new AuthorConfigCsvParser(AUTHOR_CONFIG_EMPTY_LOCATION_FILE);
+        List<RepoConfiguration> authorConfigs = authorConfigCsvParser.parse();
+        RepoConfiguration authorConfig = authorConfigs.get(0);
+
+        Assert.assertEquals(1, authorConfigs.size());
+        Assert.assertEquals(expectedConfig, authorConfig);
+        Assert.assertEquals(expectedConfig.getLocation(), authorConfig.getLocation());
+        Assert.assertEquals(expectedConfig.getBranch(), authorConfig.getBranch());
```
Comment (yong24s) [inline src/test/java/reposense/parser/CsvParserTest.java]: `Assert.assertEquals(expectedConfig.getLocation(), authorConfig.getLocation());` and `Assert.assertEquals(expectedConfig.getBranch(), authorConfig.getBranch());` are redundant since `Assert.assertEquals(expectedConfig, authorConfig);` already verified them.

File: src/test/java/reposense/parser/CsvParserTest.java [hunk 21]
```diff
@@ -139,6 +162,71 @@ public void merge_twoRepoConfigs_success() throws ParseException, IOException {
...
+        Assert.assertEquals(2, actualConfigs.size());
+        Assert.assertEquals(expectedConfigs, actualConfigs);
+
+        RepoConfiguration firstExpectedConfig = expectedConfigs.get(0);
+        RepoConfiguration secondExpectedConfig = expectedConfigs.get(1);
+        RepoConfiguration firstActualConfig = actualConfigs.get(0);
+        RepoConfiguration secondActualConfig = actualConfigs.get(1);
+
+        Assert.assertEquals(firstExpectedConfig.getLocation(), firstActualConfig.getLocation());
+        Assert.assertEquals(firstExpectedConfig.getAuthorList().hashCode(),
```
Comment (yong24s) [inline src/test/java/reposense/parser/CsvParserTest.java]: You don't need to call `hashCode()` right?

```
Assert.assertEquals(firstExpectedConfig.getAuthorList(), firstActualConfig.getAuthorList());
```

File: src/test/java/reposense/commits/CommitInfoAnalyzerTest.java (hunk 11 above)
Comment (yong24s) [reply src/test/java/reposense/commits/CommitInfoAnalyzerTest.java]: Is it better if we clear the state of `AuthorAliasMap` in a `@Before` method?

Comment (fzdy1914) [reply src/test/java/reposense/commits/CommitInfoAnalyzerTest.java]: I am considering removing the redundant adding of alias like `config.getAuthorAliasMap().put(EUGENE_AUTHOR_NAME, new Author(EUGENE_AUTHOR_NAME));` in the following method but keep the `clear()` only at where it is needed.

Comment (fzdy1914) [reply src/test/java/reposense/commits/CommitInfoAnalyzerTest.java]: Sorry for not looking on `GitTestTemplate`. Since it uses an `AllAliasAuthor` in config at the beginning, so maybe we can just put the `clear()`   in a @Before method, I will edit it and push it.

File: src/test/java/reposense/parser/CsvParserTest.java (hunk 21 above)
Comment (fzdy1914) [reply src/test/java/reposense/parser/CsvParserTest.java]: Yes, but the original `merge_twoRepoConfigs_success()` method also have this so I just followed, should I also remove this in that method?

**Comment by fzdy1914** [comment]
@yong24s Ready for review.

File: src/test/java/reposense/commits/CommitInfoAnalyzerTest.java (hunk 11 above)
Comment (yong24s) [reply src/test/java/reposense/commits/CommitInfoAnalyzerTest.java]: Deleting aliases is harder for code maintenance as the code author would need to know which aliases to delete.

The problem you faced was caused by the use of same method name, so it overrides parent method. 
So, you can do this 😃 

```
@Before
public void before() throws InvalidLocationException {
    super.before();
    config.getAuthorAliasMap().clear();
}
```

@yamidark
We may need to review and improve `GitTestTemplate`.

File: src/test/java/reposense/parser/CsvParserTest.java (hunk 21 above)
Comment (yong24s) [reply src/test/java/reposense/parser/CsvParserTest.java]: Yes. Please help me remove them. 👍

File: src/test/java/reposense/parser/CsvParserTest.java [hunk 22]
```diff
@@ -88,6 +95,20 @@ public void authorConfig_noSpecialCharacter_success() throws IOException, Invali
         Assert.assertEquals(AUTHOR_CONFIG_NO_SPECIAL_CHARACTER_AUTHORS, config.getAuthorList());
     }
 
+    @Test
+    public void authorConfig_emptyLocation_success() throws ParseException, IOException {
+        RepoConfiguration expectedConfig = new RepoConfiguration(new RepoLocation(""));
+
+        AuthorConfigCsvParser authorConfigCsvParser =
+                new AuthorConfigCsvParser(AUTHOR_CONFIG_EMPTY_LOCATION_FILE);
```
Comment (yong24s) [inline src/test/java/reposense/parser/CsvParserTest.java]: Unnecessary break, it can be together.

Our guideline
>Try to keep line length shorter than 110 characters (soft limit). But it is OK to exceed the limit slightly (hard limit: 120 chars). If the line exceeds the limit, use line wrapping at appropriate places of the line.

File: src/test/java/reposense/parser/CsvParserTest.java [hunk 23]
```diff
@@ -139,6 +160,66 @@ public void merge_twoRepoConfigs_success() throws ParseException, IOException {
...
+        Assert.assertEquals(firstExpectedConfig.getLocation(), firstActualConfig.getLocation());
+        Assert.assertEquals(firstExpectedConfig.getAuthorList(), firstActualConfig.getAuthorList());
+        Assert.assertEquals(firstExpectedConfig.getAuthorDisplayNameMap(), firstActualConfig.getAuthorDisplayNameMap());
+        Assert.assertEquals(firstExpectedConfig.getAuthorAliasMap(), firstActualConfig.getAuthorAliasMap());
+        Assert.assertEquals(REPO_LEVEL_GLOB_LIST, firstActualConfig.getIgnoreGlobList());
+
+        Assert.assertEquals(secondExpectedConfig.getLocation(), secondActualConfig.getLocation());
+        Assert.assertEquals(secondExpectedConfig.getAuthorList(), secondActualConfig.getAuthorList());
+        Assert.assertEquals(secondExpectedConfig.getAuthorDisplayNameMap(),
+                secondActualConfig.getAuthorDisplayNameMap());
```
Comment (yong24s) [inline src/test/java/reposense/parser/CsvParserTest.java]: Prefer higher-level breaks.

```
        Assert.assertEquals(
                secondExpectedConfig.getAuthorDisplayNameMap(), secondActualConfig.getAuthorDisplayNameMap());
```

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 24]
```diff
@@ -236,25 +240,59 @@ public void setIgnoreCommitList(List<CommitHash> ignoreCommitList) {
...
         authorList.add(author);
+        setAuthorDetails(author);
+        propagateIgnoreGlobList(author);
+    }
+
+    public void addAuthors(List<Author> authorList) {
+        for (Author author : authorList) {
+            if (containsAuthor(author)) {
+                logger.warning(String.format("Skipping author as %s already in repository %s",
+                        author.getGitId(), getDisplayName()));
```
Comment (yong24s) [inline src/main/java/reposense/model/RepoConfiguration.java]: Prefer higher-level breaks.

```
                logger.warning(String.format(
                        "Skipping author as %s already in repository %s", author.getGitId(), getDisplayName()));
```

**Comment by fzdy1914** [comment]
@reposense/devs Ready for review

**Comment by yong24s** [comment]
@fzdy1914 

You missed the changes requested in https://github.com/reposense/RepoSense/pull/441#discussion_r247186703.

Else, LGTM.

File: src/main/java/reposense/model/RepoConfiguration.java (hunk 15 above)
Comment (eugenepeh) [inline src/main/java/reposense/model/RepoConfiguration.java]: Possible to use `DEFAULT_BRANCH` instead of `""`? 
Rationale is that, in the unforeseen event that we perform a `git checkout ""`, it will throws an error.

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 25]
```diff
@@ -236,25 +240,59 @@ public void setIgnoreCommitList(List<CommitHash> ignoreCommitList) {
...
+
+        addAuthorAliases(author, author.getAuthorAliases());
+
+        setAuthorDisplayName(author, author.getDisplayName());
+    }
+
+    /**
+     * Propagate the {@code IgnoreGlobList} of {@code RepoConfiguration} to {@code author}.
+     */
+    public void propagateIgnoreGlobList(Author author) {
```
Comment (eugenepeh) [inline src/main/java/reposense/model/RepoConfiguration.java]: if its an internal use only method, perhaps use a private modifier.

File: src/test/java/reposense/parser/CsvParserTest.java [hunk 26]
```diff
@@ -26,14 +26,21 @@
...
     private static final Path REPO_CONFIG_NO_SPECIAL_CHARACTER_FILE = new File(CsvParserTest.class.getClassLoader()
             .getResource("CsvParserTest/repoconfig_noSpecialCharacter_test.csv").getFile()).toPath();
+    private static final Path AUTHOR_CONFIG_EMPTY_LOCATION_FILE = new File(CsvParserTest.class.getClassLoader()
+            .getResource("CsvParserTest/authorconfig_emptyLocation_test.csv").getFile()).toPath();
     private static final Path AUTHOR_CONFIG_NO_SPECIAL_CHARACTER_FILE = new File(CsvParserTest.class.getClassLoader()
             .getResource("CsvParserTest/authorconfig_noSpecialCharacter_test.csv").getFile()).toPath();
     private static final Path AUTHOR_CONFIG_SPECIAL_CHARACTER_FILE = new File(CsvParserTest.class.getClassLoader()
             .getResource("CsvParserTest/authorconfig_specialCharacter_test.csv").getFile()).toPath();
+    private static final Path MERGE_EMPTY_LOCATION_FOLDER = new File(CsvParserTest.class.getClassLoader()
+            .getResource("repoconfig_merge_empty_location_test").getFile()).toPath();
```
Comment (eugenepeh) [inline src/test/java/reposense/parser/CsvParserTest.java]: if these files are used only in `CsvParserTest`, please place them in the `CsvParserTest` folder as well to keep the test resources organized.

File: src/test/java/reposense/parser/CsvParserTest.java [hunk 27]
```diff
@@ -131,14 +151,72 @@ public void merge_twoRepoConfigs_success() throws ParseException, IOException {
...
+        RepoConfiguration firstExpectedConfig = expectedConfigs.get(0);
+        RepoConfiguration secondExpectedConfig = expectedConfigs.get(1);
+        RepoConfiguration firstActualConfig = actualConfigs.get(0);
+        RepoConfiguration secondActualConfig = actualConfigs.get(1);
+
+        Assert.assertEquals(firstExpectedConfig.getLocation(), firstActualConfig.getLocation());
+        Assert.assertEquals(firstExpectedConfig.getAuthorList(), firstActualConfig.getAuthorList());
+        Assert.assertEquals(firstExpectedConfig.getAuthorDisplayNameMap(), firstActualConfig.getAuthorDisplayNameMap());
+        Assert.assertEquals(firstExpectedConfig.getAuthorAliasMap(), firstActualConfig.getAuthorAliasMap());
+        Assert.assertEquals(REPO_LEVEL_GLOB_LIST, firstActualConfig.getIgnoreGlobList());
```
Comment (yamidark) [inline src/test/java/reposense/parser/CsvParserTest.java]: Actually, `TestUtil` has this helper method `compareRepoConfig` already which seems to be doing the same thing(before you changed from `hashCode`) here. Maybe try using that method instead, makes the code alot cleaner as well.

Comment (fzdy1914) [reply src/test/java/reposense/parser/CsvParserTest.java]: Should I change the `compareRepoConfig` method in `TestUtil` as well? Because it also contains unnecessary `hashcode`.

Comment (fzdy1914) [reply src/test/java/reposense/parser/CsvParserTest.java]: Or, we can leave it in this way and change all the unnecessary compare of RepoConfig in all tests in a seperate PR? Because there are many other tests that do not use `compareRepoConfig` method.

Comment (yamidark) [reply src/test/java/reposense/parser/CsvParserTest.java]: Let's change the `TestUtil#compareRepofConfig` method to what @yong24s has mentioned here, but don't need to touch all existing tests to use that, just ensure the new test methods you have implemented use that instead.

We can open a new issue for the other test methods to use the helper method.

**Comment by fzdy1914** [comment]
@reposense/devs Ready for review

**Comment by yamidark** [comment]
@fzdy1914 Remember to add the `Fix/Fixes #xxx` in your PR message, so that it will auto close the issue when we merge this.

**Comment by yamidark** [comment]
Nits to proposed commit message:
```
In the author-config.csv, each author must provide a location field
which is mapped to the repository to be analyzed in repo-config.csv.

However, the user may want to analyze the contribution of an author for
all the repositories in repo-config.csv, and it is inconvenient for the
user to type it out for each repository.

Let's allow users to leave the location field empty in
author-config.csv, which will add this author to all the repositories
in repo-config.csv.

Things to note, if an author with empty location exist, all the repos
will not use the default authors even there is no author listed in
author-config.csv, as this author with empty location will be added to
all the repositories.
```
